lib/categorizeAdSlot.js

```diff
@@ -0,0 +1,128 @@
+
+	// assuming if the strings are long, they can be similar but lower similarity will be detected
+	switch (averageLength) {
+		case averageLength <= 10:
+			return compareTwoStrings(tag, category) >= SHORT_STRING_SIMILARITY
+		case averageLength <= 15:
+			return compareTwoStrings(tag, category) >= MEDIUM_STRING_SIMILARITY
+		default:
+			return compareTwoStrings(tag, category) >= LONG_STRING_SIMILARITY
+	}
+}
+
+function getTagsFromCategories(categories) {
+	const categoryTitles = categories.map(c => c.toLowerCase())
+	const tags = constants.PredefinedTags.map(t => t._id)
+
+	const matchingTags = tags.filter(t => {
+		return categoryTitles.some(c => advancedMatchCheck(t, c))
+	})
+	const matches = [...categoryTitles, ...matchingTags]
+	return matches
+}
+
+function getDirNameFromCategory(category) {
+	const pathArr = category.AbsolutePath.split('/')
+	const pathWithoutLast = pathArr.slice(0, pathArr.length - 1)
+	const directoryName = pathWithoutLast.join('/')
+	return encodeURI(directoryName)
+}
+
+async function getAdditionalCategories(categories) {
+	const { CategoryData } = categories
+	const directoriesToBrowse =
+		CategoryData.constructor === Array
+			? CategoryData.map(getDirNameFromCategory)
+			: [getDirNameFromCategory(CategoryData)]
+
+	const allRequests = directoriesToBrowse.map(d => {
+		const requestURL = `https://awis.api.alexa.com/api?Action=CategoryBrowse&Output=json&Descriptions=False&Path=${d}&ResponseGroup=Categories,RelatedCategories`
+		return fetch(requestURL, {
+			headers: { 'x-api-key': AWIS_KEY },
+		})
+	})
+	const responseData = await Promise.all(allRequests)
+	const responseDataParsed = await Promise.all(responseData.map(r => r.json()))
+	const additionalCategories = responseDataParsed.map(d => {
+		const categoryObj =
+			d.Awis.Results.Result.Alexa.CategoryBrowse.Categories.Category
+		const categories =
+			categoryObj.constructor === Array
+				? categoryObj.map(c => c.Title)
+				: categoryObj.Title
+
+		const relatedCategories =
+			categoryObj.constructor === Array
+				? categoryObj.map(c => c.Title)
+				: categoryObj.Title
+		const allCategories = [...categories, ...relatedCategories]
+		return allCategories
+	})
+	const flattenedCategories = additionalCategories.reduce((all, c) => {
+		return all.concat(c)
+	})
+	return flattenedCategories
+}
+
+// Only removes duplicates currently but can be useful to blacklist tags in the future
+function cleanUpTags(tags) {
+	return tags.filter((tag, i) => tags.indexOf(tag) === i)
+}
+
+async function categorizeAdSlot(adSlot) {
+	const { website } = adSlot
+	const initialCategories = await getCategoriesFromAlexa(website)
+	if (!initialCategories) {
+		// if website isn't categorized by Alexa further checks will cause a crash
+		return []
+	}
+	const allCategories = await getAdditionalCategories(initialCategories)
+	const tags = getTagsFromCategories(allCategories)
+	const cleanTags = cleanUpTags(tags)
+	return cleanTags
```

Comment:
It will be good to put this actions in try/catch and return `[]` in case of error. It's is better than error with slot.